activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -401,12 +401,11 @@ def foreign_keys(table_name)
                  rc.delete_rule AS 'on_delete'
           FROM information_schema.referential_constraints rc
           JOIN information_schema.key_column_usage fk
-          USING (constraint_schema, constraint_name)
+          USING (constraint_schema, constraint_name,  table_name)
           WHERE fk.referenced_column_name IS NOT NULL
             AND fk.table_schema = #{scope[:schema]}
             AND fk.table_name = #{scope[:name]}
             AND rc.constraint_schema = #{scope[:schema]}
-            AND rc.table_name = #{scope[:name]}
```

Comment:
Can you provide the result of EXPLAIN for the query?

This `rc.table_name` condition is required to avoid all databases scan on `information_schema.referential_constraints`.

#27580

https://dev.mysql.com/doc/refman/5.7/en/information-schema-optimization.html

Reply:
Ah, interesting!

This is the explain before the change:

```
mysql> explain SELECT fk.referenced_table_name AS 'to_table',
    ->        fk.referenced_column_name AS 'primary_key',
    ->        fk.column_name AS 'column',
    ->        fk.constraint_name AS 'name',
    ->        rc.update_rule AS 'on_update',
    ->        rc.delete_rule AS 'on_delete'
    -> FROM information_schema.referential_constraints rc
    -> JOIN information_schema.key_column_usage fk
    -> USING (constraint_schema, constraint_name)
    -> WHERE fk.referenced_column_name IS NOT NULL
    ->   AND fk.table_schema = database()
    ->   AND fk.table_name = 'billable_owner_details'
    ->   AND rc.constraint_schema = database()
    ->   AND rc.table_name = 'billable_owner_details';
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+------------------------------------------------------------------------------------------+
| id | select_type | table | partitions | type | possible_keys | key                          | key_len | ref  | rows | filtered | Extra                                                                                    |
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+------------------------------------------------------------------------------------------+
|  1 | SIMPLE      | rc    | NULL       | ALL  | NULL          | CONSTRAINT_SCHEMA,TABLE_NAME | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned 0 databases                                        |
|  1 | SIMPLE      | fk    | NULL       | ALL  | NULL          | TABLE_SCHEMA,TABLE_NAME      | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned 0 databases; Using join buffer (Block Nested Loop) |
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+------------------------------------------------------------------------------------------+
2 rows in set, 1 warning (0.00 sec)
```

And this is the explain after the change:

```
mysql> explain SELECT fk.referenced_table_name AS 'to_table',
    ->        fk.referenced_column_name AS 'primary_key',
    ->        fk.column_name AS 'column',
    ->        fk.constraint_name AS 'name',
    ->        rc.update_rule AS 'on_update',
    ->        rc.delete_rule AS 'on_delete'
    -> FROM information_schema.referential_constraints rc
    -> JOIN information_schema.key_column_usage fk
    -> USING (constraint_schema, constraint_name, table_name)
    -> WHERE fk.referenced_column_name IS NOT NULL
    ->   AND fk.table_schema = database()
    ->   AND fk.table_name = 'billable_owner_details'
    ->   AND rc.constraint_schema = database();
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+-----------------------------------------------------------------------------------------+
| id | select_type | table | partitions | type | possible_keys | key                          | key_len | ref  | rows | filtered | Extra                                                                                   |
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+-----------------------------------------------------------------------------------------+
|  1 | SIMPLE      | rc    | NULL       | ALL  | NULL          | CONSTRAINT_SCHEMA,TABLE_NAME | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned 0 databases                                       |
|  1 | SIMPLE      | fk    | NULL       | ALL  | NULL          | TABLE_SCHEMA                 | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned 1 database; Using join buffer (Block Nested Loop) |
+----+-------------+-------+------------+------+---------------+------------------------------+---------+------+------+----------+-----------------------------------------------------------------------------------------+
2 rows in set, 1 warning (0.00 sec)
```

I was assuming that the index usage wouldn't change, but that seems to have been a flawed assumption 😕